File: spine-csharp/src/PhysicsConstraint.cs
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated April 5, 2025. Replaces all prior versions.
 *
 * Copyright (c) 2013-2026, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

using System;

namespace Spine {

	/// <summary>
	/// Applies physics to a bone.
	/// <para>
	/// See <a href="http://esotericsoftware.com/spine-physics-constraints">Physics constraints</a> in the Spine User Guide.</para>
	/// </summary>
	public class PhysicsConstraint : Constraint<PhysicsConstraint, PhysicsConstraintData, PhysicsConstraintPose> {
		internal BonePose bone;

		bool reset = true;
		float ux, uy, cx, cy, tx, ty;
		float xOffset, xLag, xVelocity;
		float yOffset, yLag, yVelocity;
		float rotateOffset, rotateLag, rotateVelocity;
		float scaleOffset, scaleLag, scaleVelocity;

		float remaining, lastTime;

		public PhysicsConstraint (PhysicsConstraintData data, Skeleton skeleton)
			: base(data, new PhysicsConstraintPose(), new PhysicsConstraintPose()) {
			if (skeleton == null) throw new ArgumentNullException("skeleton", "skeleton cannot be null.");

			bone = skeleton.bones.Items[data.bone.index].constrainedPose;
		}

		override public IConstraint Copy (Skeleton skeleton) {
			var copy = new PhysicsConstraint(data, skeleton);
			copy.pose.Set(pose);
			return copy;
		}

		/// <summary>Resets all physics state that was the result of previous movement. Use this after moving a bone to prevent physics
		/// from reacting to the movement.</summary>
		public void Reset (Skeleton skeleton) {
			remaining = 0;
			lastTime = skeleton.time;
			reset = true;
			xOffset = 0;
			xLag = 0;
			xVelocity = 0;
			yOffset = 0;
			yLag = 0;
			yVelocity = 0;
			rotateOffset = 0;
			rotateLag = 0;
			rotateVelocity = 0;
			scaleOffset = 0;
			scaleLag = 0;
			scaleVelocity = 0;
		}

		/// <summary>
		/// Translates the physics constraint so the next <see cref="Update(Skeleton, Physics)"/> forces are applied as if the bone moved an
		/// additional amount in world space.
		/// </summary>
		public void Translate (float x, float y) {
			ux -= x;
			uy -= y;
			cx -= x;
			cy -= y;
		}

		/// <summary>
		/// Rotates the physics constraint so the next <see cref="Update(Skeleton, Physics)"/> forces are applied as if the bone rotated around
		/// the specified point in world space.
		/// </summary>
		public void Rotate (float x, float y, float degrees) {
			float r = degrees * MathUtils.DegRad, cos = (float)Math.Cos(r), sin = (float)Math.Sin(r);
			float dx = cx - x, dy = cy - y;
			Translate(dx * cos - dy * sin - dx, dx * sin + dy * cos - dy);
		}

		/// <summary>Applies the constraint to the constrained bones.</summary>
		override public void Update (Skeleton skeleton, Physics physics) {
			PhysicsConstraintPose p = appliedPose;
			float mix = p.mix;
			if (mix == 0) return;

			bool x = data.x > 0, y = data.y > 0, rotateOrShearX = data.rotate > 0 || data.shearX > 0, scaleX = data.scaleX > 0;
			BonePose bone = this.bone;
			float l = bone.bone.data.length, t = data.step, z = 0;

			switch (physics) {
			case Physics.None:
				return;
			case Physics.Reset:
				Reset(skeleton);
				goto case Physics.Update; // Fall through.
			case Physics.Update:
				float delta = Math.Max(skeleton.time - lastTime, 0), aa = remaining;
				remaining += delta;
				lastTime = skeleton.time;

				float bx = bone.worldX, by = bone.worldY;
				if (reset) {
					reset = false;
					ux = bx;
					uy = by;
				} else {
					float a = remaining, i = p.inertia, f = skeleton.data.referenceScale, d = -1, m = 0, e = 0, qx = data.limit * delta,
						qy = qx * Math.Abs(skeleton.ScaleY);
					qx *= Math.Abs(skeleton.ScaleX);

					if (x || y) {
						if (x) {
							float u = (ux - bx) * i;
							xOffset += u > qx ? qx : u < -qx ? -qx : u;
							ux = bx;
						}
						if (y) {
							float u = (uy - by) * i;
							yOffset += u > qy ? qy : u < -qy ? -qy : u;
							uy = by;
						}
						if (a >= t) {
							float xs = xOffset, ys = yOffset;
							d = (float)Math.Pow(p.damping, 60 * t);
							m = t * p.massInverse;
							e = p.strength;
							float w = f * p.wind, g = f * p.gravity;
							float ax = (w * skeleton.windX + g * skeleton.gravityX) * skeleton.scaleX;
							float ay = (w * skeleton.windY + g * skeleton.gravityY) * skeleton.ScaleY;
							do {
								if (x) {
									xVelocity += (ax - xOffset * e) * m;
									xOffset += xVelocity * t;
									xVelocity *= d;
								}
								if (y) {
									yVelocity -= (ay + yOffset * e) * m;
									yOffset += yVelocity * t;
									yVelocity *= d;
								}
								a -= t;
							} while (a >= t);
							xLag = xOffset - xs;
							yLag = yOffset - ys;
						}
						z = Math.Max(0, 1 - a / t);
						if (x) bone.worldX += (xOffset - xLag * z) * mix * data.x;
						if (y) bone.worldY += (yOffset - yLag * z) * mix * data.y;
					}
					if (rotateOrShearX || scaleX) {
						float ca = (float)Math.Atan2(bone.c, bone.a), c, s, mr = 0, dx = cx - bone.worldX, dy = cy - bone.worldY;
						if (dx > qx)
							dx = qx;
						else if (dx < -qx)
							dx = -qx;
						if (dy > qy)
							dy = qy;
						else if (dy < -qy)
							dy = -qy;
						if (rotateOrShearX) {
							mr = (data.rotate + data.shearX) * mix;
							z = rotateLag * Math.Max(0, 1 - aa / t);
							float r = (float)Math.Atan2(dy + ty, dx + tx) - ca - (rotateOffset - z) * mr;
							rotateOffset += (r - (float)Math.Ceiling(r * MathUtils.InvPI2 - 0.5f) * MathUtils.PI2) * i;
							r = (rotateOffset - z) * mr + ca;
							c = (float)Math.Cos(r);
							s = (float)Math.Sin(r);
							if (scaleX) {
								r = l * bone.WorldScaleX;
								if (r > 0) scaleOffset += (dx * c + dy * s) * i / r;
							}
						} else {
							c = (float)Math.Cos(ca);
							s = (float)Math.Sin(ca);
							float r = l * bone.WorldScaleX - scaleLag * Math.Max(0, 1 - aa / t);
							if (r > 0) scaleOffset += (dx * c + dy * s) * i / r;
						}
						a = remaining;
						if (a >= t) {
							if (d == -1) {
								d = (float)Math.Pow(p.damping, 60 * t);
								m = t * p.massInverse;
								e = p.strength;
							}
							float ax = p.wind * skeleton.windX + p.gravity * skeleton.gravityX;
							float ay = p.wind * skeleton.windY + p.gravity * skeleton.gravityY;
							float rs = rotateOffset, ss = scaleOffset, h = l / f;
							if (Spine.Bone.yDown) ay = -ay;
							while (true) {
								a -= t;
								if (scaleX) {
									scaleVelocity += (ax * c - ay * s - scaleOffset * e) * m;
									scaleOffset += scaleVelocity * t;
									scaleVelocity *= d;
								}
								if (rotateOrShearX) {
									rotateVelocity -= ((ax * s + ay * c) * h + rotateOffset * e) * m;
									rotateOffset += rotateVelocity * t;
									rotateVelocity *= d;
									if (a < t) break;
									float r = rotateOffset * mr + ca;
									c = (float)Math.Cos(r);
									s = (float)Math.Sin(r);
								} else if (a < t) //
									break;
							}
							rotateLag = rotateOffset - rs;
							scaleLag = scaleOffset - ss;
						}
						z = Math.Max(0, 1 - a / t);
					}
					remaining = a;
				}
				cx = bone.worldX;
				cy = bone.worldY;
				break;
			case Physics.Pose:
				z = Math.Max(0, 1 - remaining / t);
				if (x) bone.worldX += (xOffset - xLag * z) * mix * data.x;
				if (y) bone.worldY += (yOffset - yLag * z) * mix * data.y;
				break;
			}

			if (rotateOrShearX) {
				float o = (rotateOffset - rotateLag * z) * mix, s, c, a;
				if (data.shearX > 0) {
					float r = 0;
					if (data.rotate > 0) {
						r = o * data.rotate;
						s = (float)Math.Sin(r);
						c = (float)Math.Cos(r);
						a = bone.b;
						bone.b = c * a - s * bone.d;
						bone.d = s * a + c * bone.d;
					}
					r += o * data.shearX;
					s = (float)Math.Sin(r);
					c = (float)Math.Cos(r);
					a = bone.a;
					bone.a = c * a - s * bone.c;
					bone.c = s * a + c * bone.c;
				} else {
					o *= data.rotate;
					s = (float)Math.Sin(o);
					c = (float)Math.Cos(o);
					a = bone.a;
					bone.a = c * a - s * bone.c;
					bone.c = s * a + c * bone.c;
					a = bone.b;
					bone.b = c * a - s * bone.d;
					bone.d = s * a + c * bone.d;
				}
			}
			if (scaleX) {
				float s = 1 + (scaleOffset - scaleLag * z) * mix * data.scaleX;
				bone.a *= s;
				bone.c *= s;
			}
			if (physics != Physics.Pose) {
				tx = l * bone.a;
				ty = l * bone.c;
			}
			bone.ModifyWorld(skeleton.update);
		}

		override public void Sort (Skeleton skeleton) {
			Bone bone = this.bone.bone;
			skeleton.SortBone(bone);
			skeleton.updateCache.Add(this);
			skeleton.SortReset(bone.children);
			skeleton.Constrained(bone);
		}

		override public bool IsSourceActive { get { return bone.bone.active; } }

		/// <summary>The bone constrained by this physics constraint.</summary>
		public BonePose Bone { get { return bone; } set { bone = value; } }
	}
}

File: spine-cpp/src/spine/PhysicsConstraint.cpp
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated April 5, 2025. Replaces all prior versions.
 *
 * Copyright (c) 2013-2025, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

#include <spine/PhysicsConstraint.h>
#include <spine/PhysicsConstraintData.h>
#include <spine/PhysicsConstraintPose.h>
#include <spine/BonePose.h>
#include <spine/Skeleton.h>
#include <spine/SkeletonData.h>
#include <spine/BoneData.h>
#include <spine/Bone.h>
#include <spine/MathUtil.h>

using namespace spine;

RTTI_IMPL(PhysicsConstraint, Constraint)

PhysicsConstraint::PhysicsConstraint(PhysicsConstraintData &data, Skeleton &skeleton)
	: PhysicsConstraintBase(data), _reset(true), _ux(0), _uy(0), _cx(0), _cy(0), _tx(0), _ty(0), _xOffset(0), _xLag(0), _xVelocity(0), _yOffset(0),
	  _yLag(0), _yVelocity(0), _rotateOffset(0), _rotateLag(0), _rotateVelocity(0), _scaleOffset(0), _scaleLag(0), _scaleVelocity(0), _remaining(0),
	  _lastTime(0) {

	_bone = &skeleton._bones[(size_t) data._bone->getIndex()]->_constrainedPose;
}

PhysicsConstraint &PhysicsConstraint::copy(Skeleton &skeleton) {
	PhysicsConstraint *copy = new (__FILE__, __LINE__) PhysicsConstraint(_data, skeleton);
	copy->_pose.set(_pose);
	return *copy;
}

void PhysicsConstraint::reset(Skeleton &skeleton) {
	_remaining = 0;
	_lastTime = skeleton.getTime();
	_reset = true;
	_xOffset = 0;
	_xLag = 0;
	_xVelocity = 0;
	_yOffset = 0;
	_yLag = 0;
	_yVelocity = 0;
	_rotateOffset = 0;
	_rotateLag = 0;
	_rotateVelocity = 0;
	_scaleOffset = 0;
	_scaleLag = 0;
	_scaleVelocity = 0;
}

void PhysicsConstraint::translate(float x, float y) {
	_ux -= x;
	_uy -= y;
	_cx -= x;
	_cy -= y;
}

void PhysicsConstraint::rotate(float x, float y, float degrees) {
	float r = degrees * MathUtil::Deg_Rad, cosVal = MathUtil::cos(r), sinVal = MathUtil::sin(r);
	float dx = _cx - x, dy = _cy - y;
	translate(dx * cosVal - dy * sinVal - dx, dx * sinVal + dy * cosVal - dy);
}

void PhysicsConstraint::update(Skeleton &skeleton, Physics physics) {
	PhysicsConstraintPose &p = *_appliedPose;
	float mix = p._mix;
	if (mix == 0) return;

	bool x = _data._x > 0, y = _data._y > 0, rotateOrShearX = _data._rotate > 0 || _data._shearX > 0, scaleX = _data._scaleX > 0;
	BonePose *bone = _bone;
	float l = bone->_bone->_data.getLength(), t = _data._step, z = 0;

	switch (physics) {
		case Physics_None:
			return;
		case Physics_Reset:
			reset(skeleton);
			// Fall through.
		case Physics_Update: {
			float delta = MathUtil::max(skeleton._time - _lastTime, 0.0f), aa = _remaining;
			_remaining += delta;
			_lastTime = skeleton._time;

			float bx = bone->_worldX, by = bone->_worldY;
			if (_reset) {
				_reset = false;
				_ux = bx;
				_uy = by;
			} else {
				float a = _remaining, i = p._inertia, f = skeleton._data.getReferenceScale(), d = -1, m = 0, e = 0, qx = _data._limit * delta,
					  qy = qx * MathUtil::abs(skeleton.getScaleY());
				qx *= MathUtil::abs(skeleton._scaleX);
				if (x || y) {
					if (x) {
						float u = (_ux - bx) * i;
						_xOffset += u > qx ? qx : u < -qx ? -qx : u;
						_ux = bx;
					}
					if (y) {
						float u = (_uy - by) * i;
						_yOffset += u > qy ? qy : u < -qy ? -qy : u;
						_uy = by;
					}
					if (a >= t) {
						float xs = _xOffset, ys = _yOffset;
						d = MathUtil::pow(p._damping, 60 * t);
						m = t * p._massInverse;
						e = p._strength;
						float w = f * p._wind, g = f * p._gravity;
						float ax = (w * skeleton._windX + g * skeleton._gravityX) * skeleton._scaleX;
						float ay = (w * skeleton._windY + g * skeleton._gravityY) * skeleton.getScaleY();
						do {
							if (x) {
								_xVelocity += (ax - _xOffset * e) * m;
								_xOffset += _xVelocity * t;
								_xVelocity *= d;
							}
							if (y) {
								_yVelocity -= (ay + _yOffset * e) * m;
								_yOffset += _yVelocity * t;
								_yVelocity *= d;
							}
							a -= t;
						} while (a >= t);
						_xLag = _xOffset - xs;
						_yLag = _yOffset - ys;
					}
					z = MathUtil::max(0.0f, 1 - a / t);
					if (x) bone->_worldX += (_xOffset - _xLag * z) * mix * _data._x;
					if (y) bone->_worldY += (_yOffset - _yLag * z) * mix * _data._y;
				}
				if (rotateOrShearX || scaleX) {
					float ca = MathUtil::atan2(bone->_c, bone->_a), c, s, mr = 0, dx = _cx - bone->_worldX, dy = _cy - bone->_worldY;
					if (dx > qx)
						dx = qx;
					else if (dx < -qx)
						dx = -qx;
					if (dy > qy)
						dy = qy;
					else if (dy < -qy)
						dy = -qy;
					if (rotateOrShearX) {
						mr = (_data._rotate + _data._shearX) * mix;
						z = _rotateLag * MathUtil::max(0.0f, 1 - aa / t);
						float r = MathUtil::atan2(dy + _ty, dx + _tx) - ca - (_rotateOffset - z) * mr;
						_rotateOffset += (r - MathUtil::ceil(r * MathUtil::InvPi_2 - 0.5f) * MathUtil::Pi_2) * i;
						r = (_rotateOffset - z) * mr + ca;
						c = MathUtil::cos(r);
						s = MathUtil::sin(r);
						if (scaleX) {
							r = l * bone->getWorldScaleX();
							if (r > 0) _scaleOffset += (dx * c + dy * s) * i / r;
						}
					} else {
						c = MathUtil::cos(ca);
						s = MathUtil::sin(ca);
						float r = l * bone->getWorldScaleX() - _scaleLag * MathUtil::max(0.0f, 1 - aa / t);
						if (r > 0) _scaleOffset += (dx * c + dy * s) * i / r;
					}
					a = _remaining;
					if (a >= t) {
						if (d == -1) {
							d = MathUtil::pow(p._damping, 60 * t);
							m = t * p._massInverse;
							e = p._strength;
						}
						float ax = p._wind * skeleton._windX + p._gravity * skeleton._gravityX;
						float ay = p._wind * skeleton._windY + p._gravity * skeleton._gravityY;
						float rs = _rotateOffset, ss = _scaleOffset, h = l / f;
						while (true) {
							a -= t;
							if (scaleX) {
								_scaleVelocity += (ax * c - ay * s - _scaleOffset * e) * m;
								_scaleOffset += _scaleVelocity * t;
								_scaleVelocity *= d;
							}
							if (rotateOrShearX) {
								_rotateVelocity -= ((ax * s + ay * c) * h + _rotateOffset * e) * m;
								_rotateOffset += _rotateVelocity * t;
								_rotateVelocity *= d;
								if (a < t) break;
								float r = _rotateOffset * mr + ca;
								c = MathUtil::cos(r);
								s = MathUtil::sin(r);
							} else if (a < t)
								break;
						}
						_rotateLag = _rotateOffset - rs;
						_scaleLag = _scaleOffset - ss;
					}
					z = MathUtil::max(0.0f, 1 - a / t);
				}
				_remaining = a;
			}
			_cx = bone->_worldX;
			_cy = bone->_worldY;
			break;
		}
		case Physics_Pose: {
			z = MathUtil::max(0.0f, 1 - _remaining / t);
			if (x) bone->_worldX += (_xOffset - _xLag * z) * mix * _data._x;
			if (y) bone->_worldY += (_yOffset - _yLag * z) * mix * _data._y;
			break;
		}
	}

	if (rotateOrShearX) {
		float o = (_rotateOffset - _rotateLag * z) * mix, s, c, a;
		if (_data._shearX > 0) {
			float r = 0;
			if (_data._rotate > 0) {
				r = o * _data._rotate;
				s = MathUtil::sin(r);
				c = MathUtil::cos(r);
				a = bone->_b;
				bone->_b = c * a - s * bone->_d;
				bone->_d = s * a + c * bone->_d;
			}
			r += o * _data._shearX;
			s = MathUtil::sin(r);
			c = MathUtil::cos(r);
			a = bone->_a;
			bone->_a = c * a - s * bone->_c;
			bone->_c = s * a + c * bone->_c;
		} else {
			o *= _data._rotate;
			s = MathUtil::sin(o);
			c = MathUtil::cos(o);
			a = bone->_a;
			bone->_a = c * a - s * bone->_c;
			bone->_c = s * a + c * bone->_c;
			a = bone->_b;
			bone->_b = c * a - s * bone->_d;
			bone->_d = s * a + c * bone->_d;
		}
	}
	if (scaleX) {
		float s = 1 + (_scaleOffset - _scaleLag * z) * mix * _data._scaleX;
		bone->_a *= s;
		bone->_c *= s;
		switch (_data._scaleYMode) {
			case ScaleYMode_Uniform:
				bone->_b *= s;
				bone->_d *= s;
				break;
			case ScaleYMode_Volume:
				s = MathUtil::abs(s);
				s = s >= 0.7f ? 1 / s : 4 - 3.67347f * s;
				bone->_b *= s;
				bone->_d *= s;
				break;
			default:
				break;
		}
	}
	if (physics != Physics_Pose) {
		_tx = l * bone->_a;
		_ty = l * bone->_c;
	}
	bone->modifyWorld(skeleton._update);
}

void PhysicsConstraint::sort(Skeleton &skeleton) {
	Bone *bone = _bone->_bone;
	skeleton.sortBone(bone);
	skeleton._updateCache.add(this);
	skeleton.sortReset(bone->_children);
	skeleton.constrained(*bone);
}

bool PhysicsConstraint::isSourceActive() {
	return _bone->_bone->isActive();
}

BonePose &PhysicsConstraint::getBone() {
	return *_bone;
}

void PhysicsConstraint::setBone(BonePose &bone) {
	_bone = &bone;
}
